src/state/store.ts

```diff
@@ -168,70 +152,70 @@ export const useRecipeStore = create<RecipeStore>()(
-            set((s) => {
-                const rec = s.recipes[recipeId];
-                if (!rec) return s;
-                return {
-                    recipes: {
-                        ...s.recipes,
-                        [recipeId]: {
-                            ...rec,
-                            currentString: newString,
-                            isModified: newString !== rec.originalString,
-                        },
-                    },
-                };
-            });
-        },
 
-        updateRecipeObj: (recipeId, updates) => {
+        editRecipe: (recipeId, path, value) => {
             const rec = get().recipes[recipeId];
             if (!rec) return;
 
-            try {
-                const obj = JSON.parse(rec.currentString);
+            const newEdits = { ...rec.edits };
 
-                for (const [path, value] of Object.entries(updates)) {
-                    lodashSet(obj, path, value);
-                }
-                get().updateRecipeString(
-                    recipeId,
-                    JSON.stringify(obj, null, 2)
-                );
-            } catch {
-                // TODO: better error handling
-                console.warn("Failed to update recipe object");
+            const defaultValue = lodashGet(rec.defaultRecipeData, path);
+            if (isEqual(defaultValue, value)) {
+                delete newEdits[path]; // no longer different from default
+            } else {
+                newEdits[path] = value;
             }
-        },
 
-        restoreRecipeDefault: (recipeId) => {
-            const rec = get().recipes[recipeId];
-            if (rec) get().updateRecipeString(recipeId, rec.originalString);
+            set((state) => ({
+                recipes: {
+                    ...state.recipes,
+                    [recipeId]: {
+                        ...rec,
+                        edits: newEdits
+                    },
+                },
+            }));
         },
 
+
         getCurrentValue: (path) => {
             const { selectedRecipeId, recipes } = get();
-            const str = recipes[selectedRecipeId]?.currentString;
-            if (!str) return undefined;
-            try {
-                const obj = JSON.parse(str);
-                const v = lodashGet(obj, path);
-                return typeof v === "string" || typeof v === "number"
-                    ? v
-                    : undefined;
-            } catch {
-                console.warn("Failed to retrieve value.");
+            const rec = recipes[selectedRecipeId];
+            if (!rec) return undefined;
+
+            // First check if an edited value exists at this path
+            const editedValue = lodashGet(rec.edits, path);
+            if (editedValue !== undefined) {
+                if (typeof editedValue === "string" || typeof editedValue === "number") {
+                    return editedValue;
+                }
                 return undefined;
```

Comment:
I can't think of a time where we'd actually hit this `return undefined;` case, so perhaps it doesn't really matter, but I feel like we would want to try to continue on to trying to get the default value rather than just returning undefined?